File: src/utils/RateLimiter.cpp

#include "RateLimiter.hpp"

namespace LinkEmbed {

RateLimiter::RateLimiter(double rate_per_second) : rate(rate_per_second) {
    allowance = rate;
    last_check = std::chrono::steady_clock::now();
}

bool RateLimiter::TryAcquire() {
    std::lock_guard<std::mutex> lock(mutex);
    auto now = std::chrono::steady_clock::now();
    double time_passed = std::chrono::duration<double>(now - last_check).count();
    last_check = now;

    allowance += time_passed * rate;
    if (allowance > rate) {
        allowance = rate;
    }

    if (allowance >= 1.0) {
        allowance -= 1.0;
        return true;
    }
    return false;
}

}

File: src/rate_limiter.rs
use std::sync::Mutex;
use std::time::Instant;

pub struct RateLimiter {
    rate: f64,
    state: Mutex<State>,
}

struct State {
    allowance: f64,
    last_check: Instant,
}

impl RateLimiter {
    pub fn new(rate: f64) -> Self {
        Self {
            rate,
            state: Mutex::new(State {
                allowance: rate,
                last_check: Instant::now(),
            }),
        }
    }

    pub fn try_acquire(&self) -> bool {
        let mut state = match self.state.lock() {
            Ok(guard) => guard,
            Err(_) => return false,
        };

        let now = Instant::now();
        let elapsed = now.duration_since(state.last_check).as_secs_f64();
        state.last_check = now;

        state.allowance += elapsed * self.rate;
        if state.allowance > self.rate {
            state.allowance = self.rate;
        }

        if state.allowance >= 1.0 {
            state.allowance -= 1.0;
            true
        } else {
            false
        }
    }
}

#[cfg(test)]
mod tests {
    use std::thread;
    use std::time::Duration;

    use super::RateLimiter;

    #[test]
    fn enforces_basic_token_bucket_behavior() {
        let limiter = RateLimiter::new(2.0);
        assert!(limiter.try_acquire());
        assert!(limiter.try_acquire());
        assert!(!limiter.try_acquire());
        thread::sleep(Duration::from_millis(600));
        assert!(limiter.try_acquire());
    }
}
